src/detect.c

```diff
@@ -2299,9 +2299,16 @@ static int SignatureCreateMask(Signature *s)
         SCLogDebug("sig requires http app state");
     }
 
-    if (s->sm_lists[DETECT_SM_LIST_HSBDMATCH] != NULL) {
-        s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
-        SCLogDebug("sig requires http app state");
+    if (s->sm_lists[DETECT_SM_LIST_FILEDATA] != NULL) {
+        if (s->alproto == ALPROTO_HTTP)
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
+        else if (s->alproto == ALPROTO_SMTP)
+            s->mask |= SIG_MASK_REQUIRE_SMTP_STATE;
+        else if (s->alproto == ALPROTO_UNKNOWN) {
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
```

Comment:
This is wrong. This would cause this rule not to be inspected unless it has both a http and a smtp state. This will never happen. In this case you can just not set these flags.